lib/types/btrfs.nix

```diff
@@ -1,4 +1,52 @@
 { config, options, diskoLib, lib, rootMountPoint, parent, device, ... }:
+let
+  swapType = lib.types.submodule ({ config, ... }: {
+    options = {
+      enable = lib.mkEnableOption "swap";
+
+      size = lib.mkOption {
+        type = lib.types.nullOr (lib.types.strMatching "^([0-9]+[KMGTP])?$");
+        default = null;
+        description = "Size of the swap file (e.g. 2G)";
+      };
+
+      path = lib.mkOption {
+        type = lib.types.str;
+        default = "swapfile";
+        description = "Path to the swap file (relative to the mountpoint)";
+      };
+    };
+  });
+
+  swapConfig = { mountpoint, swap }: lib.optional swap.enable {
+    swapDevices = [{
+      device =
+        assert lib.asserts.assertMsg (mountpoint != null) "swap requires mountpoint to be set";
+        "${mountpoint}/${swap.path}";
+    }];
+  };
+
+  subvolSwapCreate = mountpoint: swap:
+    if !swap.enable
+    then ""
+    else
+      assert lib.asserts.assertMsg (swap.size != null) "swap size must be set";
+      ''btrfs filesystem mkswapfile --size ${swap.size} ${mountpoint}/${swap.path}'';
+
+  partitionSwapCreate = device: swap:
+    if !swap.enable
+    then ""
+    else
+      assert lib.asserts.assertMsg (swap.size != null) "swap size must be set";
+      ''
+        (
+          MNTPOINT=$(mktemp -d)
+          mount ${device} "$MNTPOINT" -o subvol=/
+          trap 'umount $MNTPOINT; rm -rf $MNTPOINT' EXIT
```

Comment:
The trap is scoped here to the subshell right?

Reply:
I'm not a bash expert - but I copied this from further down [in the same file](https://github.com/nix-community/disko/blob/b32f0e2/lib/types/btrfs.nix#L82). So if it's wrong, it's wrong for all of the subvolume creations as well.

Reply:
Sounds about right.